src/cephadm/design_choices.rst

```diff
@@ -0,0 +1,165 @@
+
+There are a couple of solutions to this:
+
+
+1. Pre fetching/caching data on the host
+----------------------------------------
+
+This would allow bin/cephadm to autonomously scrape data that can later be retrieved by mgr/cephadm.
+
+i.e. `cephadm ls` would access an underlying (transparent) cache.
+
+This, btw assumes a **pull** model.
+
+There is at least one possible solutions on *how* to do this.
+
+1.1 daemonize bin/cephadm
+--------------------------
+
+A daemon is able to execute scraping tasks more frequently (as opposed to only when asked by mgr/cephadm)
+
+
+Downsides:
+
+We're still `scp'ing` bin/cephadm from the ceph-mgr to the host which is then executed.
+A daemon on the other hand needs to run continuously on the target system. If ceph-mgr (and the containing bin/cephadm
+script) is updated, the daemon needs to be informed of this event and get updated/restarted as well.
+
+
+1.1.1 HTTP:
+___________
+
+There is a implementation by Paul Cuzner [2] that uses HTTP to serve an endpoint to the hosts metadata.
+
+Upside:
+
+* We could utilize a proper REST API to CRUD all of the daemons/disks etc.
+  Currently it feels a bit weird to call a binary from a remote host (that's another topic though)
+
+
+Downside:
+
+* This is quite some shift in architecture.
+* We still use SSH to bootstrap a node (in order to install bin/cephadm for the first time)
+  Having two distinct transport layers feels odd.
+* The current approach of delivering bin/cephadm to the host doesn't allow the use of external dependencies.
+  This means that we're stuck with the built-in HTTP server lib, which isn't meant for production purposes.
+  bin/cephadm needs to be packaged and distributed (one way or the other) for us to make use of a production ready
+  http server.
+
+Requirements:
+
+* https
+* authentication
+* documentation
+* api versioning (keep it stable)
+
+
+1.1.2 Keep bin/cephadm
+______________________
+
+Daemonizing bin/cephadm allows us to pre-fetch data and store it in some shape or form.
+
+Currently we're only interested in speeding up metadata retrieval. This could mean that we don't even
+need to introduce a new transport layer but rather transparently return cached content via the `cephadm ls`
+or the `cephadm ceph-volume ls` command.
+
+The technical details need to be figured out. (format of data storage etc.)
+
+Upside:
+
+* No new transport layer.
+  Also means no extra auth, security layer is needed
+
+Downside:
+
+* May feel a bit hacky
+* Needs locks (if a refresh is currently happening)
+
+
+1.1.3 Use k-v store
+___________________
+
+After the mgr/cephadm queried metadata from each host, we usually save it in the mon's k-v store.
+
+If each host would be allowed to write their own metadata to the store, we'd not need to explicitly call
+for a refresh.
+
+This needs research on some topics:
+
+* Is the mon's k-v store capable of accepting **many** requests (CRUD)
```

Comment:
This all comes down to my initial question on what scenario we are aiming at... 100's, thousand nodes...?

I don't know ceph-mon KV storage internals, but I've had a really bad time in the past trying to use SSTable-based storage (Cassandra) for time series (or datasets often and periodically updated).

Reply:
@jecluis What's your take on this?

Reply:
We've had our fair share of issues over the years with the frequency and size of updates to the mon store -- it is great for serving reads, but not so nice if a lot of data is coming in all the time.

I think there are a few things that need to be considered here:
* size of updates
* frequency of updates
* amount of nodes updating

So far, the monitors have been, for the vast majority of cluster participants, read-only: they serve maps, auth tickets, and not much more. Now we are considering having dozens, hundreds, or thousands of nodes adding information directly to the monitors, right?

Personally, I think that is quite a bad idea. The monitors should not be seen as a scalable, general purpose persistent k/v store. It was not designed with that purpose in mind, and is sharing the same commit path as mission critical updates such as maps. This has been fine because updates to the kv store have not put pressure on map updates so far, and competition has been minimal. What we are proposing is having all hosts competing for proposals on a k/v store that was not designed to handle that kind of traffic.

Furthermore, the monitors are mission critical. We should not want meddle with it unless the information kept is mission critical as well. So far we've had a lot of exceptions on this, but, as I said, it's been small amounts of (mostly) read-only information that is essential for cluster operation.

I have not read through this proposal, but I have a feeling that keeping metadata about hosts may not fall under the absolutely mission critical requirement. This sort of information may make more sense to keep in rados.

Reply:
unfortunately, cephadm needs actually both:

* It needs a persistent storage that provides the performance to handle lots of meta data updates. 
* It needs a persistent storage in a cluster only containing a single MON+MGR in order to deploy OSDs. 

And right now the config-key store is the only thing that is usable for us.